### PR TITLE
feat: add filter by column type on repositories screen

### LINKED ISSUE #627: feature: Filter by column type on Repository screen
### Type of feature

Select the type of feature request, the lowercase should also be the PR prefix.

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

### Current behavior

Is your feature request related to a problem? Please describe.

When viewing the Repository page, it has been requesting users, the ability to filter by activity, pr count, etc.  

### Suggested solution

When the user clicks the column the data should sort ascending or descending. 

Screen Shot 2022-11-30 at 4 27 14 PM

### Code of Conduct

By submitting this issue, you agree to follow our [Code of Conduct](https://docs.opensauced.pizza/contributing/code-of-conduct/).

- [ ] I agree to follow this project's Code of Conduct

### Contributing Docs

If you plan on contributing code please read - [Contribution Guide](https://docs.opensauced.pizza/contributing/introduction-to-contributing/)

- [ ]  I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by TAKANOME-DEV**
## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Description

This PR adds the ability to filter the data in ascending or descending order when the user clicks the column.
Here is how I did it:
- add a `click handler` to the column that will do the following:
  - set the  `orderBy` state to the column name
  - set  the `orderDirection` to `ASC` or `DESC` based on the previous value
  - add both the `orderBy` and `orderDirection` to the url
- create a custom hook that takes `orderBy`,`orderDirection` and the repos data, and return the sorted data
- add  the `Sorted by` to the `FilterHeader` :arrow_down: 
filter
- and finally, clear the sorting when the `X`  button is clicked
## Related Tickets & Documents

Fixes #627 
## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [x] 📕 storybook
- [ ] 🙅 no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?
unstoppable

### CODE REVIEW

**Comment by TAKANOME-DEV** [comment]
@bdougie right now, we can only sort repos by `stars`(default),  `name` and `owner`. If we have to hit the server to get the sorted data, then this PR is ready. Clicking the table header named `Repository` will sort the repos by their name in `asc` or `desc` order.

IMO, that's not really helpful because when you change pages and sort the data, you will only get different repos sorted by their name in `asc` or `desc` order (it's hard to explain :sweat_smile: but try it out in the deploy preview by filtering the repos, changing pages and filter them one more time :eyes: )

I would suggest to only filter the data (note that it's 10 per page) on the client side by their name, activity, pr count etc.

What's your thoughts about it?

**Comment by bdougie** [comment]
> IMO, that's not really helpful because when you change pages and sort the data, you will only get different repos sorted by their name in asc or desc order (it's hard to explain 😅 but try it out in the deploy preview by filtering the repos, changing pages and filter them one more time 👀 )

Maybe record a loom and share.

If there are 100 repos, we should sort once and pagination should respect that sorting.

As mentioned we can simply sort the current page for now. Users can expand the list to 100 if they would like to see more. They also have the option to filter above for the most active.

**Comment by TAKANOME-DEV** [comment]
@bdougie we can sort the data by name, pr count, pr velocity and spam. Note that they are done inside `Repositories organisms`(where we have access to the table title).
However, I didn't manage to sort them by activity or contribution because those data are fetched using different hooks (by providing the repo `id`) inside `RepoRow` molecule (repositories -> repository-table -> repo-row).

**Comment by bdougie** [comment]
Ok, we should keep the scope focused. 

## Activity

Clicking Activity should add the "the most active filter." To be clear, activity is a filter and not a sort. We can add a "least active filter" in the API.  

Screen Shot 2022-12-16 at 5 13 01 PM

## Sorting

- For now, sort only what we can for now. 
- The arrow should not show unless the items are sorted
- A new `sorted by` should be shown when sorting is present and not shown when no sorting.  If sorted by name, velocity, etc. It should show and be removed

Screen Shot 2022-12-16 at 5 17 42 PM

File: components/molecules/SortedBySelector/sorted-by-selector.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  handleCancelClick: () => void;
+  className?: string;
+  selected?: string;
+}
+
+const SortedOrders = {
+  "name": "Repository",
+  "prsCount": "Pull requests count",
+  "prVelocityCount": "Pull requests velocity",
+  "spamPrsCount": "Spam pull requests count"
```
Comment (bdougie) [inline components/molecules/SortedBySelector/sorted-by-selector.tsx:20]:
```suggestion
  "prsCount": "PR Count",
  "prVelocityCount": "PR Velocity",
  "spamPrsCount": "Spam"
```

**Review by bdougie — CHANGES_REQUESTED** [review]
These names should math the column names. 

I also see the sorting is a bugging. The repository filter never sorts ascending and the only filter that does consistently is PR Overview.  

https://www.loom.com/share/89458bb5bd834e73b1d954346fb33732

File: components/molecules/SortedBySelector/sorted-by-selector.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  "prVelocityCount": "PR Velocity",
+  "spamPrsCount": "Spam"
+};
+
+const SortedBySelector: React.FC<SortedBySelectorProps> = ({
+  handleCancelClick,
+  selected
+}) => {
+
+  if (!selected) return null;
```
Comment (brandonroberts) [inline components/molecules/SortedBySelector/sorted-by-selector.tsx:28]: This must return an element, even if its empty

```suggestion
  if (!selected) return <></>;
```

**Review by brandonroberts — CHANGES_REQUESTED** [review]

File: components/organisms/Repositories/repositories.tsx [hunk 3]
```diff
@@ -14,15 +15,25 @@ import useSupabaseAuth from "lib/hooks/useSupabaseAuth";
...
+
+type FilterOptions = keyof DbRepo | "prsCount";
+
 const Repositories = ({ repositories }: RepositoriesProps): JSX.Element => {
+  const [orderBy, setOrderBy] = useState<FilterOptions | null>(null);
+  const [orderDirection, setOrderDirection] = useState("");
   const { user } = useSupabaseAuth();
   const router = useRouter();
-  const { filterName, toolName, selectedFilter, userOrg } = router.query;
+  const { filterName, toolName, selectedFilter, userOrg, sortedBy } = router.query;
```
Comment (brandonroberts) [inline components/organisms/Repositories/repositories.tsx:36]:
```suggestion
  const { filterName, toolName, selectedFilter, userOrg, orderBy } = router.query;
```

**Review by brandonroberts — CHANGES_REQUESTED** [review]

**Comment by brandonroberts** [comment]
@TAKANOME-DEV when I click on the repository column header, I still see `sortedBy` in the URL: https://deploy-preview-662--oss-insights.netlify.app/javascript/repositories/filter/recent?sortedBy=name&sort=ASC.

Also not sure what its sorting by for this list

image

**Comment by TAKANOME-DEV** [comment]
> @TAKANOME-DEV when I click on the repository column header, I still see `sortedBy` in the URL: https://deploy-preview-662--oss-insights.netlify.app/javascript/repositories/filter/recent?sortedBy=name&sort=ASC.
> 
Hum, I did change it though. Let me double check that.

> Also not sure what its sorting by for this list
> 
It's sorting by the repository name

**Comment by TAKANOME-DEV** [comment]
@brandonroberts you mean something like this :arrow_down: 

| Before | After |
| --------- | --------- |
| before | after |

**Comment by brandonroberts** [comment]
@TAKANOME-DEV not lowercasing the names themselves, but lowercasing them when you're sorting

**Comment by TAKANOME-DEV** [comment]
> @TAKANOME-DEV not lowercasing the names themselves, but lowercasing them when you're sorting

@brandonroberts I did it in my last commit https://github.com/open-sauced/insights/pull/662/commits/b834a2850d3f1e0d4f058e177660b8d6dbe48fe9

**Review by brandonroberts — APPROVED** [review]
LGTM

Sorting is working as intended now

**Comment by bdougie** [comment]
I'd rather prioritize the use experience over the code quality. We can refactor in code review.

**Comment by brandonroberts** [comment]
@TAKANOME-DEV I would say build the URL based on what's available and push a new navigation after everything is checked

**Comment by TAKANOME-DEV** [comment]
> @TAKANOME-DEV I would say build the URL based on what's available and push a new navigation after everything is checked

@brandonroberts could you expand on that? I'm not sure what you mean 🤔

**Comment by brandonroberts** [comment]
Sure. All the currently applied filters are from the current URL, which could from the `router.query` object.

The `topic/filterName` should always be present, (javascript)
The `toolName` should always be present - (repositories)
If the `selectedFilter` is not empty, use it (filter/{selectedFilter) or (filter/{repoFullName})

So it should be

```
/{topic}/{toolName}/{selectedFilter-if-present}
```

**Comment by bdougie** [comment]
recording

Also recorded a bug where a user can remove all filters and sorting (by default) and see an error when attempting to sort again. 

<details><summary>Browser metadata</summary>





```
Path:      /javascript/repositories/filter/undefined?orderBy=name&sort=ASC
Browser:   Chrome 108.0.0.0 on Mac OS 10.15.7
Viewport:  1146 x 1016 @1x
Language:  en-US
Cookies:   Enabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-662--oss-insights.netlify.app%2Fjavascript%2Frepositories%2Ffilter%2Fundefined%3ForderBy%3Dname%26sort%3DASC&os=OS%20X&os_version=Catalina&browser=Chrome&browser_version=108.0&resolution=1146x1016&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-662--oss-insights.netlify.app/javascript/repositories/filter/undefined?orderBy=name&sort=ASC) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=63a0ec67d206d10008be35ad&commentID=63a1f1a80255c40f65d3671c&resolution=resolved)

**Comment by TAKANOME-DEV** [comment]
@bdougie it should be okay now, re-deploy and test it :+1:

**Comment by bdougie** [comment]
Looks good from here 

recording


<details><summary>Browser metadata</summary>





```
Path:      /javascript/repositories/filter/recent?orderBy=prsCount&sort=ASC
Browser:   Chrome 108.0.0.0 on Mac OS 10.15.7
Viewport:  1440 x 796 @2x
Language:  en-US
Cookies:   Enabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-662--oss-insights.netlify.app%2Fjavascript%2Frepositories%2Ffilter%2Frecent%3ForderBy%3DprsCount%26sort%3DASC&os=OS%20X&os_version=Catalina&browser=Chrome&browser_version=108.0&resolution=1440x796&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-662--oss-insights.netlify.app/javascript/repositories/filter/recent?orderBy=prsCount&sort=ASC) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=63a1f411ce9c8b000ab189fe&commentID=63a264f4763ee2076b632bed&resolution=resolved)

**Review by bdougie — APPROVED** [review]
QA'd the preview and it looks good.

File: components/organisms/Repositories/repositories.tsx [hunk 4]
```diff
@@ -10,19 +11,32 @@ import TableHeader from "components/molecules/TableHeader/table-header";
...
-import useStore from "lib/store";
 
 interface RepositoriesProps {
   repositories?: number[];
 }
 
+type FilterOptions = keyof DbRepo | "prsCount" |"activity"
+
+const renderArrow = (order: string | undefined) => {
+  return order === "ASC" ? <FaArrowUp className="text-light-slate-11 ml-2" fontSize={16} />
```
Comment (NsdHSO) [inline components/organisms/Repositories/repositories.tsx:27]: can you please create an enum `... enum { ASC="ASC"}`

File: lib/hooks/useFilterRepos.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+      setFilteredRepos(sortedRepos);
+    }
+  }, [repos, filterName, orderDirection]);
+
+  return { filteredRepos };
+};
+
+export default useFilterRepos;
+
+function isDbRepo(repo: any): repo is DbRepo {
```
Comment (NsdHSO) [inline lib/hooks/useFilterRepos.ts:38]: can you please avoid to use `any`

Comment (TAKANOME-DEV) [reply lib/hooks/useFilterRepos.ts:38]: @NsdHSO  this function is a `custom type guard`, it's only used to assert that the typeof `a` and `b` is `DbRepo`
```ts
// line 17
 if (isDbRepo(a) && isDbRepo(b)) {
  // now the typeof `a` and `b` is narrowed and we can access all the methods of `DbRepo`
}
```

File: components/organisms/Repositories/repositories.tsx (hunk 4 above)
Comment (TAKANOME-DEV) [reply components/organisms/Repositories/repositories.tsx:27]: Could you expand on that? 
I'm not sure I understand

Comment (NsdHSO) [reply components/organisms/Repositories/repositories.tsx:27]: `enum Order {
  ASC= 'ASC',
  DESC="DESC"
} `
And here you have to change in 
`return order === Order.ASC ? <FaArrowUp className="text-light-slate-11 ml-2" fontSize={16} />`

**Pull request merged**